src/main/java/fi/aalto/cs/apluscourses/model/ExerciseDataSource.java

```diff
@@ -0,0 +1,25 @@
+package fi.aalto.cs.apluscourses.model;
+
+import java.io.IOException;
+import java.util.List;
+import org.jetbrains.annotations.NotNull;
+
+public interface ExerciseDataSource {
```

Comment:
so is it so, that the `ExerciseDataSource` is an **"Exercise"** and the `Exercise` is a **"ExerciseImpl"**?

Reply:
No. `ExerciseDataSource` is an object that is used when communicating with the API. Basically, it is encapsulation of the API whose methods are called to get model objects.

Why do we need it? It is beneficial that the API code is in one place. Abstracting it behind an interface allows us to implement model classes (and everything else) separately from the API (currently not fully done because JSON parsing is still in the model classes but that should be changed). Now, if for example a new (non-backwards-compatible) version of API is published we can just write a new implementation of `ExerciseDataSource` to comply that. The choice between API version could then be dynamically by just using an object of desired class.